extra/tensors/tensors.factor

```diff
@@ -0,0 +1,257 @@
+! Copyright (C) 2019 HMC Clinic.
+! See http://factorcode.org/license.txt for BSD license.
+USING: accessors alien.c-types alien.data arrays
+concurrency.combinators grouping kernel locals math.functions
+math.ranges math multi-methods quotations sequences sequences.private
+specialized-arrays tensors.tensor-slice typed ;
+QUALIFIED-WITH: alien.c-types c
+SPECIALIZED-ARRAY: c:float
+IN: tensors
+
+! Tensor class definition
+TUPLE: tensor
+    { shape array }
+    { vec float-array } ;
+
+! Errors
+ERROR: shape-mismatch-error shape1 shape2 ;
+
+<PRIVATE
+
+! Construct a tensor of zeros
+: <tensor> ( shape seq -- tensor )
```

Comment:
May want some basic assertions that the shape has positive (or at least non-negative?) values.

Reply:
This has now been included in the functions that call `<tensor>`, because it sometimes needs to happen well before `<tensor>` is called.